File: drills/elixir/solutions/nc104_word_dictionary__by_length.exs
defmodule Solution do
  # Words bucketed by length. A pattern can only match words of its own length,
  # so that one check throws away most of the collection before any character is
  # compared.
  def new, do: %{}

  def add_word(store, word) do
    Map.update(store, String.length(word), [word], &[word | &1])
  end

  # No shared prefixes, so every candidate of the right length is compared
  # position by position. Slower than the trie on a large dictionary, and it
  # needs no tree -- which is the trade the trie is making.
  def search(store, word) do
    pattern = String.graphemes(word)

    store
    |> Map.get(String.length(word), [])
    |> Enum.any?(fn candidate ->
      pattern
      |> Enum.zip(String.graphemes(candidate))
      |> Enum.all?(fn {p, c} -> p == "." or p == c end)
    end)
  end
end

File: drills/go/solutions/nc104_word_dictionary__by_length.go
package main

// Words bucketed by length: a query only has to be compared against the
// words of its own length, position by position with dots as wildcards.
type WordDictionary struct {
	byLength map[int][]string
}

func Constructor() WordDictionary {
	return WordDictionary{byLength: map[int][]string{}}
}

func (d *WordDictionary) AddWord(word string) {
	d.byLength[len(word)] = append(d.byLength[len(word)], word)
}

func (d *WordDictionary) Search(word string) bool {
	for _, candidate := range d.byLength[len(word)] {
		matches := true
		for i := 0; i < len(word); i++ {
			if word[i] != '.' && word[i] != candidate[i] {
				matches = false
				break
			}
		}
		if matches {
			return true
		}
	}
	return false
}
